### ISSUE TITLE
Iterator / stream consumer can not use subclasses of iterator/stream

### ISSUE THREAD

**Issue by hgschmie** [jdbi/jdbi#2210]
See the AttachedHandleLeakTest in the sqlobject module:

```java
public interface UserDao extends SqlObject {
        @SqlQuery("SELECT * from users order by id")
        void getIterableUsers(Consumer<ResultIterator<User>> consumer);
}
```

```
    @Test
    public void testLeakIteratorConsumer() {
        for (int i = 0; i < 1000; i++) {
            dao.getIterableUsers(it -> assertThat(it.next()).extracting("id").isEqualTo(1));
        }
    }
```

```
java.lang.ClassCastException: class java.util.Spliterators$1Adapter cannot be cast to class org.jdbi.v3.core.result.ResultIterator (java.util.Spliterators$1Adapter is in module java.base of loader 'bootstrap'; org.jdbi.v3.core.result.ResultIterator is in unnamed module of loader 'app')

	at org.jdbi.v3.sqlobject.statement.internal.ResultReturner$ConsumeIteratorResultReturner.accept(ResultReturner.java:377)
	at org.jdbi.v3.sqlobject.statement.internal.ResultReturner$ConsumerResultReturner.mappedResult(ResultReturner.java:333)
	at org.jdbi.v3.sqlobject.statement.internal.ResultReturner$ConsumerResultReturner.mappedResult(ResultReturner.java:303)
	at org.jdbi.v3.sqlobject.statement.internal.SqlQueryHandler.lambda$configureReturner$0(SqlQueryHandler.java:61)
	at org.jdbi.v3.sqlobject.statement.internal.CustomizingStatementHandler.invoke(CustomizingStatementHandler.java:196)
	at org.jdbi.v3.sqlobject.statement.internal.SqlQueryHandler.invoke(SqlQueryHandler.java:27)
	at org.jdbi.v3.sqlobject.internal.SqlObjectInitData$1.lambda$invoke$0(SqlObjectInitData.java:135)
	at org.jdbi.v3.core.ConstantHandleSupplier.invokeInContext(ConstantHandleSupplier.java:62)
	at org.jdbi.v3.sqlobject.internal.SqlObjectInitData$1.call(SqlObjectInitData.java:141)
	at org.jdbi.v3.sqlobject.internal.SqlObjectInitData$1.invoke(SqlObjectInitData.java:135)
	at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$attach$2(SqlObjectFactory.java:107)
	at jdk.proxy2/jdk.proxy2.$Proxy18.getIterableUsers(Unknown Source)
	at org.jdbi.v3.sqlobject.AttachedHandleLeakTest.testLeakIteratorConsumer(AttachedHandleLeakTest.java:107)
[...]
```

Which is sensible, given that the returned iterator from a stream is an iterator, not a ResultIterator. The culprits are most likely  https://github.com/jdbi/jdbi/blob/master/sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java#L320 and https://github.com/jdbi/jdbi/blob/master/sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java#L323; a similar piece of code here : https://github.com/jdbi/jdbi/blob/master/sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java#L75-L80 does exact matches for Stream.class and Iterator.class. Also, it does explicit checking for ResultIterator

**Comment by stevenschlansker** [comment]
This wasn't really intended to work. Since you are explicitly delegating the resource ownership to Jdbi, there seems to be no reason to use a ResultIterator, which mostly exists to combine Iterator and Closeable.

Is there a particular use case you can think of where this would be valuable to do?

**Comment by hgschmie** [comment]
I am not disagreeing; what I would have expected is a "xxx not supported" exception, not an obscure classcast exception.

**Comment by hgschmie** [comment]
Fixed by #2217
